usr/man/man4/rx.4
.\" Copyright (c) 1980 Regents of the University of California.
.\" All rights reserved.  The Berkeley software License Agreement
.\" specifies the terms and conditions for redistribution.
.\"
.\"	@(#)rx.4	6.2 (Berkeley) 8/20/87
.\"
.TH RX 4 "August 20, 1987"
.UC 2
.SH NAME
rx \- DEC RX02 floppy disk
.SH SYNOPSIS
.ft B
.nf
/sys/conf/SYSTEM:
	NRX	\fIrx_units\fP	# RX02

/etc/dtab:
.ta .5i +\w'#Name 'u +\w'Unit# 'u +\w'177777 'u +\w'Vector 'u +\w'Br 'u +\w'xxxxxxx 'u +\w'xxxxxxx 'u
	#Name	Unit#	Addr	Vector	Br	Handler(s)		# Comments
	rx	?	177170	264	5	rxintr		# rx02
.DT

major device number(s):
	raw: 17
	block: 8
minor device encoding:
	bit 01 specifies drive: <rx_drive>
	bit 02 specifies density: single: 0, double: 1
.fi
.ft R
.SH DESCRIPTION
The
.I rx
device provides access to a DEC RX02 floppy disk
unit with M8256 interface module (RX211 configuration). 
The RX02 uses 8-inch, single-sided, soft-sectored floppy
disks (with pre-formatted industry-standard headers) in
either single or double density.
.PP
Floppy disks handled by the RX02 contain 77 tracks, each with 26
sectors (for a total of 2,002 sectors).  The sector size is 128
bytes for single density, 256 bytes for double density.  Single 
density disks are compatible with the RX01 floppy disk unit and with
IBM 3740 Series Diskette 1 systems.  
.PP
Files with minor device numbers 0 and 1 refer to drives 0 and 1
in single density mode;
minor devices 2 and 3 refer to drives 0 and 1
in double density mode.
The standard device names begin with ``rx'' followed by
the drive number and then a letter a-b for single and double density
access to the drive respectively.
The character ? stands here for a drive number in the range 0-1.
.PP
The block files access the disk via the system's normal
buffering mechanism and may be read and written without regard to
physical disk records.  There is also a `raw' interface
which provides for direct transmission between the disk
and the user's read or write buffer.
A single read or write call results in exactly one I/O operation
and therefore raw I/O is considerably more efficient when
many words are transmitted.  The names of the raw files
conventionally begin with an extra `r.'
.PP
In raw I/O the buffer must begin on a word (even) boundary,
and counts should be a multiple of the disk's sector size
(either 128 or 256 bytes depending on selected density.)
Likewise
.I seek
calls should specify a multiple of the disk's sector size.
.PP
In addition to normal (`block' and `raw') i/o, the driver supports
formatting of disks for either density.
.SH "DISK SUPPORT"
The
.I rx
driver does not support pseudo-disks (partitions).  Each file
.I rx?[ab]
refers to the entire drive as a single sequentially addressed
file.  The physical disk sector size is 128 bytes for single
density and 256 bytes for double density.  The logical block
size is 1024 bytes.  Each diskette has 250 logical blocks in
single density and 500 logical blocks in double density.
.PP
The size and density of the disks are specified as follows:
.PP
.nf
.ta .5i +\w'rx00  'u +\w'minor device  'u +\w'unit  'u
	disk	minor device	unit	density
	rx0a	0	0	single
	rx1a	1	1	single
	rx0b	2	0	double
	rx1b	3	1	double
.DT
.fi
.PP
Even though the storage capacity on a floppy disk is quite
small, it is possible to make file systems on them.
For example, the commands
.nf
.RS
% rxformat /dev/rrx0a
% newfs /dev/rrx0a
.RE
and
.RS
% rxformat /dev/rrx1b
% newfs /dev/rrx1b
.RE
.fi
format and make file systems on the single density disk in the left drive
and the double density disk in the right drive with
241 and 486 kbytes available, respectively, for file storage.
Using
.IR tar (1)
gives somewhat more efficient utilization of the available
space for file storage.
The RX02 disks are primarily useful for small archives and transfer of
small systems or distributions.
.PP
An
.IR ioctl (2)
call is available is available to format RX02 disks:
.RS
.nf
.ft B
#include <pdpuba/rxreg.h>
ioctl(fildes, RXIOC_FORMAT)
.ft R
.fi
.RE
The density is determined by the device opened.
.SH FILES
.ta \w'/dev/MAKEDEV.local  'u
/dev/rx[01][ab]	block files
.br
/dev/rrx[01][ab]	raw files
.br
/dev/MAKEDEV	script to create special files
.br
/dev/MAKEDEV.local	script to localize special files
.DT
.SH "SEE ALSO"
hk(4),
ra(4),
ram(4),
rk(4),
rl(4),
si(4),
xp(4),
dtab(5),
autoconfig(8),
rxformat(8V)
.SH DIAGNOSTICS
\fBrx2%d: hard error sn%d cs=%b er=%b\fR.  An unrecoverable
error occurred during transfer of the specified sector of the specified
disk.  The contents of the two error registers are also printed
in octal and symbolically with bits decoded.
The error was either unrecoverable, or a large number of retry attempts
could not recover the error.
.SH BUGS
In raw I/O
.I read
and
.IR write (2)
truncate file offsets to disk sector size block boundaries
(either 128 or 256 bytes depending on the selected density),
and
.I write
scribbles on the tail of incomplete blocks.
Thus,
in programs that are likely to access raw devices,
.I read, write
and
.IR lseek (2)
should always deal in disk sector size multiples.
.PP
DEC-standard error logging should be supported.
.PP
A program to analyze the logged error information (even in its
present reduced form) is needed.
.PP
The 4.3BSD rx driver which supports far more capabilities should be ported
to 2.10BSD.


usr/src/sys/OTHERS/rx02/#2/rx.c
/*
 * Copyright (c) 1986 Regents of the University of California.
 * All rights reserved.  The Berkeley software License Agreement
 * specifies the terms and conditions for redistribution.
 *
 *	@(#)rx.c	1.2 (2.11BSD GTE) 1/2/93
 */

/*
 * RX02 floppy disk device driver
 *
 * This driver was written by Bill Shannon and distributed on the
 * DEC v7m UNIX tape.  It has been modified for 2BSD and has been
 * included with the permission of the DEC UNIX Engineering Group.
 *
 * Modified to actually work with 2.9BSD by Gregory Travis, Indiana Univ.
 *
 *	Layout of logical devices:
 *
 *	name	min dev		unit	density
 *	----	-------		----	-------
 *	rx0a	   0		  0	single
 *	rx1a	   1		  1	single
 *	rx0b	   2		  0	double
 *	rx1b	   3		  1	double
 *
 *	ioctl function call may be used to format a disk.
 */

#include "rx.h"
#if NRX > 0
#include "param.h"
#include "buf.h"
#include "conf.h"
#include "ioctl.h"
#include "tty.h"
#include "rxreg.h"
#include "errno.h"

struct	rxdevice *RXADDR;

/*
 *	the following defines use some fundamental
 *	constants of the RX02.
 */
#define	NSPB	((minor(bp->b_dev)&2) ? 2 : 4)		/* sectors per block */
#define	NRXBLKS	((minor(bp->b_dev)&2) ? 1001 : 500)	/* blocks on device */
#define	NBPS	((minor(bp->b_dev)&2) ? 256 : 128)	/* bytes per sector */
#define	DENSITY	(minor(bp->b_dev)&2)	/* Density: 0 = single, 2 = double */
#define	UNIT	(minor(bp->b_dev)&1)	/* Unit Number: 0 = left, 1 = right */

#define	rxwait()	while (((RXADDR->rxcs) & RX_XREQ) == 0)
#define	seccnt(bp)	((int)((bp)->b_seccnt))

struct	buf	rxtab;
struct	buf	crxbuf;		/* buffer header for control functions */

/*
 *	states of driver, kept in b_state
 */
#define	SREAD	1	/* read started */
#define	SEMPTY	2	/* empty started */
#define	SFILL	3	/* fill started */
#define	SWRITE	4	/* write started */
#define	SINIT	5	/* init started */
#define	SFORMAT	6	/* format started */

rxattach(addr, unit)
	struct rxdevice *addr;
	u_int unit;
{
    if (unit != 0)
	return (0);
    RXADDR = addr;
    return (1);
}

/*ARGSUSED*/
rxopen(dev, flag)
	dev_t dev;
{
	if (minor(dev) >= 4 || !RXADDR)
		return (ENXIO);
	return (0);
}

rxstrategy(bp)
	register struct buf *bp;
{
	register int s;

	if (minor(bp->b_dev) >= 4 || !RXADDR)
		goto bad;
	mapalloc(bp);
	if (bp->b_blkno >= NRXBLKS) {
		if (bp->b_flags&B_READ)
			bp->b_resid = bp->b_bcount;
		else {
bad:			bp->b_flags |= B_ERROR;
			bp->b_error = ENXIO;
		}
		iodone(bp);
		return;
	}
	bp->av_forw = (struct buf *) NULL;

	/*
	 * seccnt is actually the number of floppy sectors transferred,
	 * incremented by one after each successful transfer of a sector.
	 */
	seccnt(bp) = 0;

	/*
	 * We'll modify b_resid as each piece of the transfer
	 * successfully completes.  It will also tell us when
	 * the transfer is complete.
	 */
	bp->b_resid = bp->b_bcount;
	s = splbio();
	if (rxtab.b_actf == NULL)
		rxtab.b_actf = bp;
	else
		rxtab.b_actl->av_forw = bp;
	rxtab.b_actl = bp;
	if (rxtab.b_state == NULL)
		rxstart();
	splx(s);
}

rxstart()
{
	register struct buf *bp;
	int addr, xmem;
	int n, sector, track;

	if ((bp = rxtab.b_actf) == NULL) {
		rxtab.b_state = NULL;
		return;
	}

	if (bp == &crxbuf) {		/* is it a control request ? */
		rxtab.b_state = SFORMAT;
		RXADDR->rxcs = RX_SMD | RX_GO | RX_IE
					| (UNIT << 4) | (DENSITY << 7);
		rxwait();
		RXADDR->rxdb = 0222;
	} else
	if (bp->b_flags & B_READ) {
		rxtab.b_state = SREAD;
		rxfactr((int)bp->b_blkno * NSPB + seccnt(bp), &sector, &track);
		RXADDR->rxcs = RX_RSECT | RX_GO | RX_IE
					| (UNIT << 4) | (DENSITY << 7);
		rxwait();
		RXADDR->rxsa = sector;
		rxwait();
		RXADDR->rxta = track;
	} else {
		rxtab.b_state = SFILL;
		n = bp->b_resid >= NBPS ? NBPS : bp->b_resid;
		rxaddr ( bp, &addr, &xmem );
		RXADDR->rxcs = RX_FILL | RX_Q22 | RX_GO
					| RX_IE | (DENSITY << 7);
		rxwait();
		RXADDR->rxwc = n >> 1;
		rxwait();
		RXADDR->rxba = addr;
		rxwait();
		RXADDR->rxba = xmem;
	}
}

rxintr()
{
	register struct buf *bp;
	int n, sector, track;
static	rxerr[4];
	char	*decode;
	int addr, xmem;

	if (rxtab.b_state == SINIT) {
		rxstart();
		return;
	}

	if ((bp = rxtab.b_actf) == NULL)
		return;

	if (RXADDR->rxcs & RX_ERR) {
		if (rxtab.b_errcnt++ > 10 || rxtab.b_state == SFORMAT) {
			bp->b_flags |= B_ERROR;
			harderr(bp, "rx");
			printf("cs=%b er=%b\n", RXADDR->rxcs, RX_BITS,
				RXADDR->rxes, RXES_BITS);
			RXADDR->rxcs = RX_RDEC | RX_GO;
			rxwait();
			RXADDR->rxba = (short) rxerr;
			while ( ! (RXADDR->rxcs & RX_DONE) );
			switch ( rxerr[0] )
			{
			case 0040: decode = "bad track"; break;
			case 0050: decode = "found home"; break;
			case 0070: decode = "no sch sctr"; break;
			case 0120: decode = "no preamble"; break;
			case 0150: decode = "ozone headers"; break;
			case 0160: decode = "too many IDAM"; break;
			case 0170: decode = "data AM missing"; break;
			case 0200: decode = "CRC error"; break;
			case 0240: decode = "density error"; break;
			case 0250: decode = "bad fmt key"; break;
			case 0260: decode = "bad data AM"; break;
			case 0270: decode = "POK while write"; break;
			case 0300: decode = "drv not ready"; break;
			case 0310: decode = "write protected"; break;
			default: decode = "unknown error"; break;
			}
			printf ( "rx: errcode %o = %s\n", rxerr[0], decode );
			rxtab.b_errcnt = 0;
			rxtab.b_actf = bp->av_forw;
			iodone(bp);
		}
		RXADDR->rxcs = RX_INIT;
		RXADDR->rxcs = RX_IE;
		rxtab.b_state = SINIT;
		return;
	}
	switch (rxtab.b_state) {

	case SREAD:			/* read done, start empty */
		rxtab.b_state = SEMPTY;
		n = bp->b_resid >= NBPS? NBPS : bp->b_resid;
		rxaddr ( bp, &addr, &xmem );
		RXADDR->rxcs = RX_EMPTY | RX_Q22 | RX_GO
					| RX_IE | (DENSITY << 7);
		rxwait();
		RXADDR->rxwc = n >> 1;
		rxwait();
		RXADDR->rxba = addr;
		rxwait();
		RXADDR->rxba = xmem;
		return;

	case SFILL:			/* fill done, start write */
		rxtab.b_state = SWRITE;
		rxfactr((int)bp->b_blkno * NSPB + seccnt(bp), &sector, &track);
		RXADDR->rxcs = RX_WSECT | RX_GO | RX_IE
					| (UNIT << 4) | (DENSITY << 7);
		rxwait();
		RXADDR->rxsa = sector;
		rxwait();
		RXADDR->rxta = track;
		return;

	case SEMPTY:			/* empty done, start next read */
	case SWRITE:			/* write done, start next fill */
		/*
		 * increment amount remaining to be transferred.
		 * if it becomes positive, last transfer was a
		 * partial sector and we're done, so set remaining
		 * to zero.
		 */
		if (bp->b_resid <= NBPS) {
done:
			bp->b_resid = 0;
			rxtab.b_errcnt = 0;
			rxtab.b_actf = bp->av_forw;
			iodone(bp);
			break;
		}

		bp->b_resid -= NBPS;
		seccnt(bp)++;
		break;

	case SFORMAT:			/* format done (whew!!!) */
		goto done;		/* driver's getting too big... */
	}

	/* end up here from states SWRITE and SEMPTY */
	rxstart();
}

/*
 *	rxfactr -- calculates the physical sector and physical
 *	track on the disk for a given logical sector.
 *	call:
 *		rxfactr(logical_sector,&p_sector,&p_track);
 *	the logical sector number (0 - 2001) is converted
 *	to a physical sector number (1 - 26) and a physical
 *	track number (0 - 76).
 *	the logical sectors specify physical sectors that
 *	are interleaved with a factor of 2. thus the sectors
 *	are read in the following order for increasing
 *	logical sector numbers (1,3, ... 23,25,2,4, ... 24,26)
 *	There is also a 6 sector slew between tracks.
 *	Logical sectors start at track 1, sector 1; go to
 *	track 76 and then to track 0.  Thus, for example, unix block number
 *	498 starts at track 0, sector 25 and runs thru track 0, sector 2
 *	(or 6 depending on density).
 */
static
rxfactr(sectr, psectr, ptrck)
	register int sectr;
	int *psectr, *ptrck;
{
	register int p1, p2;

	p1 = sectr / 26;
	p2 = sectr % 26;
	/* 2 to 1 interleave */
	p2 = (2 * p2 + (p2 >= 13 ?  1 : 0)) % 26;
	/* 6 sector per track slew */
	*psectr = 1 + (p2 + 6 * p1) % 26;
	if (++p1 >= 77)
		p1 = 0;
	*ptrck = p1;
}


/*
 *	rxaddr -- compute core address where next sector
 *	goes to / comes from based on bp->b_un.b_addr, bp->b_xmem,
 *	and seccnt(bp).
 */
static
rxaddr(bp, addr, xmem)
	register struct buf *bp;
	register char **addr, *xmem;
{
	*addr = bp->b_un.b_addr + seccnt(bp) * NBPS;
	*xmem = bp->b_xmem;
	if (*addr < bp->b_un.b_addr)		/* overflow, bump xmem */
		(*xmem)++;
}

/*
 *	rxioctl -- format RX02 disk, single or double density.
 *	density determined by device opened.
 */
/*ARGSUSED*/
rxioctl(dev, cmd, addr, flag)
	dev_t dev;
	u_int cmd;
{
	register int s;
	register struct buf *bp;

	if (cmd != RXIOC_FORMAT)
		return (ENXIO);
	bp = &crxbuf;
	while (bp->b_flags & B_BUSY) {
		s = splbio();
		bp->b_flags |= B_WANTED;
		sleep(bp, PRIBIO);
	}
	splx(s);
	bp->b_flags = B_BUSY;
	bp->b_dev = dev;
	bp->b_error = 0;
	rxstrategy(bp);
	iowait(bp);
	bp->b_flags = 0;
	return (0);
}
#endif
